datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -767,6 +769,71 @@ private Future<Boolean> dispatchAssignmentChangeIfNeeded(String connectorType, L
     return null;
   }
 
+  private List<DatastreamTask> getRemovedTasks(List<DatastreamTask> newAssignment, List<DatastreamTask> oldAssignment) {
+    List<DatastreamTask> removedTasks = new ArrayList<>(oldAssignment);
+    removedTasks.removeAll(newAssignment);
+    return removedTasks;
+  }
+
+  private List<DatastreamTask> getAddedTasks(List<DatastreamTask> newAssignment, List<DatastreamTask> oldAssignment) {
+    List<DatastreamTask> addedTasks = new ArrayList<>(newAssignment);
+    addedTasks.removeAll(oldAssignment);
+    return addedTasks;
+  }
+
+  @VisibleForTesting
+  void maybeClaimAssignmentTokensForStoppingStreams(List<DatastreamTask> newAssignment,
+      List<DatastreamTask> oldAssignment) {
+    Map<String, List<DatastreamTask>> newAssignmentPerConnector = new HashMap<>();
+    for (DatastreamTask task : newAssignment) {
+      String connectorType = task.getConnectorType();
+      newAssignmentPerConnector.computeIfAbsent(connectorType, k -> new ArrayList<>()).add(task);
+    }
+
+    Map<String, List<DatastreamTask>> oldAssignmentPerConnector = new HashMap<>();
+    for (DatastreamTask task : oldAssignment) {
+      String connectorType = task.getConnectorType();
+      oldAssignmentPerConnector.computeIfAbsent(connectorType, k -> new ArrayList<>()).add(task);
+    }
+
+    Set<String> allConnectors = new HashSet<>();
+    allConnectors.addAll(newAssignmentPerConnector.keySet());
+    allConnectors.addAll(oldAssignmentPerConnector.keySet());
+
+    // The follower nodes don't keep an up-to-date view of the datastreams Zk node and making them listen to changes
+    // to datastreams node risks overwhelming the ZK server with reads. Instead, the follower is inferring the stopping
+    // streams from the task assignment for each connector type. Even if that inference results in a falsely identified
+    // stopping stream, the attempt to claim the token will be a no-op.
+    for (String connector : allConnectors) {
+      List<DatastreamTask> oldTasks = oldAssignmentPerConnector.getOrDefault(connector, Collections.emptyList());
+      List<DatastreamTask> newTasks = newAssignmentPerConnector.getOrDefault(connector, Collections.emptyList());
+      List<DatastreamTask> removedTasks = getRemovedTasks(newTasks, oldTasks);
+
+      List<Datastream> stoppingStreams = inferStoppingDatastreamsFromAssignment(newTasks, removedTasks);
+      List<String> stoppingStreamNames = stoppingStreams.stream().map(Datastream::getName).collect(Collectors.toList());
+
+      if (!stoppingStreamNames.isEmpty()) {
```

Comment:
You probably need `else` from the logging/debugging aspect.

Reply:
Adding